Refactor

Weekndd opened the pull request:
1. Finish 클래스명을 Findwinner로 수정.
2. Findwinner 클래스에 있는 For문을 Stream으로 수정해서 활용.
3. Input클래스를 만들어서 생성자에서 하던 예외처리를 분리.
4. Input클래스에 checkPlayerName으로 예외처리를 담당하는 메서드 생성.
5. View클래스를 만들어서 기존의 runCount는 Car클래스에서 숫자로 관리하고 화면 출력부분을 따로 분리.
6. RacingCarTest(테스트 코드)에 5글자 이상일 시 예외처리와 FindWinner클래스의 우승자를 출력하는 테스트작성
(예외처리가 정상적으로 돌아가는지 확인하는 테스트는 예외처리가 잘 되면 예외처리가 잘된다는 메시지와 함께 예외가 터짐..이게 맞나?)

Weekndd committed:
Refactor

테스트코드를 위한 random클래스 생성,
Car클래스에서 run메서드 분리

Weekndd committed:
Refactor

미니멈값과 맥스값을 명시해서 가독성을 높임

Weekndd committed:
Delete

파일명 불일치로 재업로드

Weekndd committed:
Refactor

미니멈값과 맥스값을 명시해서 가독성을 높임

Junuu commented:
1. readme에 구현하고자 하는 게임의 규칙이 들어가면 좋을것 같음
2. Main.java로 이루어져 있어 file Changes의 가독성이 조금 떨어지는것 같음 main 클래스의 내부클래스가 아닌 실제 클래스로 분리해보기 (비슷하게 테스트 코드들도 src/java/test 테스트파일로 빼서 관리하자)
3. 클래스, 함수, 메서드의 이름은 최대한 명확하고 의미있게 (줄임말은 안쓰는게 좋다) -> 추후에 내가 다시보거나, 다른 사람이 보았을때 어떻게 느낄지 시선으로 작성하자

Junuu commented on Main.java:
```diff
@@ -0,0 +1,172 @@
+        /**
+         * 랜덤값 정의 클래스
+         */
+        class RandomNum{
+        	final int MAXNUM = 9;
+            public int generate(){
+                Random random = new Random();
+                return random.nextInt(MAXNUM);
+            }
+        }
```
RandomNumberGenerator라는 클래스이름이 랜덤의 숫자를 생성한다는 의미를 명확하게 드러낼 것 같음

Junuu commented on Main.java:
```diff
@@ -0,0 +1,172 @@
+            //플레이어의 이름이 5글자 이상이면 예외를 발생시켜서 메시지를 뜨게하고 다시 입력받는다.
+            //while문으로 처리함 (정상적인 값을 받을 때 까지)
+
+            public String[] join(String playerInput) {
+                String[] playerStr = playerInput.split(","); //쉼표를 기준으로 문자열을 나눈다.
+                while (true) {
+                    try {
+                        for (int i = 0; i < playerStr.length; i++) {
+                            if (playerStr[i].length() >= 6) {
+                                throw new IllegalArgumentException();
+                            }
+                        }
+                        break;
+                    } catch (IllegalArgumentException e) {
+                        System.out.println("[ERROR]5글자 이하의 이름을 입력하세요!");
+                        playerInput = input.next();
+                        playerStr = playerInput.split(",");  //재입력된 값을 쉼표를 기준으로 나눈다.
+                    }
+                }
+                return playerStr;
+            }
```
Player의 이름이 5글자 이상이라는 정책이 과연 입력받는곳에서 드러나는게 좋을까?
아니면 Car 클래스에서 이름을 주어 생성할때 드러나는게 좋을까?
어떻게 보면 join메서드는 다양한 역할을 수행하고 있어

1. 문자열을 나눈다.
2. 나눈 문자열이 6개이상이라면 예외를 터트린다.

단순히 문자열을 나누어 반환하도록 하면 어떨까?

Junuu commented on Main.java:
```diff
@@ -0,0 +1,172 @@
+         * 카 클래스
+         */
+        class Car{
+            String name = new String();
+            Integer runCount = 0;   //몇 번 움직였는지 숫자
+            String result = "";     //몇 번 움직였는지 문자
```
car 클래스를 생성할때 new String()이 필요한지?
runCount와 result는 어떻게 다른지? 왜 2개가 필요한지 잘 모르겠어

Junuu commented on Main.java:
```diff
@@ -0,0 +1,172 @@
+            }
+            
+            
+            public String run(int randomNum,Car player) {
+            	final int MINNUM = 4;
+                if (randomNum >= MINNUM) {
+                    player.result = player.result.concat("-");
+                    player.runCount++;
+                }
+                return player.result;
+            }
```
runCount를 직접 조작하는건 객체의 캡슐화에 조금 어긋나는것 같아
Car클래스가 직접 move메서드로 책임을 갖는건 어떨까? 이렇게되면 Car.move()를 수행할때 randomNum에 따라 움직이거나 움직이지 않거나를 추상적으로 표현할 수 있을것 같아.

추가적으로 MINIMUM같은 경우도 최솟값보다는 움직이기위한 임계값이라는 표현이 더 명확할것 같아.
ex) MOVE_THRESHOLD

Junuu commented on Main.java:
```diff
@@ -0,0 +1,172 @@
+
+
+
+            public void winner(String[] playerStr, ArrayList<Car> player) {
+                //runCount가 가장 높은 값 찾기
+                Integer winnerResult = 0;
+                for(int i=0; i<playerStr.length; i++) {
+                    if(winnerResult < player.get(i).runCount) {
+                        winnerResult = player.get(i).runCount;
+                    }
+                }
+                //runCount 가장 높은 값 출력
+                String winner = ""; //우승자 출력할 문자열
+                for(int i=0; i<playerStr.length; i++) {
+                    if(player.get(i).runCount == winnerResult) {
+                        winner = winner.concat(player.get(i).name+",");
+                    }
+                }
+                //마지막에 붙은 쉼표값 제거
+                if(winner.endsWith(",")){
+                    winner = winner.substring(0,winner.length()-1);
+                };
+                System.out.println("최종 우승자: "+winner);
+            }
```
playStr이 필요한 이유는 무엇일까?
이미 Car클래스에서 이름은 필드로 가지고 있으니 이것을 활용하여 우승자를 출력하면 되지 않을까?

Weekndd committed:
Refactor

1. readme 게임규칙으로 수정
2. 클래스들을 main내부에서 외부로 분리
3. 변수와 메서드이름을 명확하게 수정
4. 6글자 이상의 경우, join 메서드에서 하던 예외처리를 생성자에서 하도록 수정(단순히 문자열만 반환하도록 수정)
5. 중복되는 변수 및 필요 없는 변수들 삭제 및 수정
6. run메서드를 Car클래스로 옮겨서 캡슐화에 더 부합하도록 수정

Weekndd committed:
Delete

Main.java 파일삭제
새로 올린 src파일의 Main.java와 중복

junuuuuuu commented on RacingCarTest.java:
```diff
@@ -0,0 +1,21 @@
+        //Then
+        Assertions.assertEquals(1,testCar.getRunCount());
+
+        //When
+        testCar.run(1,testCar);
+        //Then
+        Assertions.assertEquals(2,testCar.getRunCount());
```
이 부분은 4이하인 숫자인 1이 들어가서 run 메서드를 실행했으므로, length가 2가되지않고 그대로 1이 되어야 하는 테스트여야 할것 같다는 생각이 들어

junuuuuuu replied:
추가적으로 다양한 테스트들이 추가되면 좋을것 같아
1. Car를 생성할때 이름이 5글자 이상이면 예외가 발생하고 그렇지 않다면 잘 생성되는지 확인하는 테스트
2. FindWinner 클래스에서 우승자를 정상적으로 잘 출력하는지 확인하는 테스트

junuuuuuu commented on src/Car.java:
```diff
@@ -0,0 +1,35 @@
+            }
+        }
+    }
+    public static String run(int randomNumber, Car player) {
+        final int MOVE_THRESHOLD = 4;
+        if (randomNumber >= MOVE_THRESHOLD ) {
+            player.runCount = player.runCount.concat("-");
+        }
+        return player.runCount;
+    }
```
runCount를 문자열로 관리하는데 아마 화면에 그리는 부분까지 고려해서 코드를 작성한것 같아보여.
자동차의 위치를 숫자로 관리하고, 화면에 그리는 부분은 View 클래스를 따로두어 거기서 보여주면 어떨까?

그러면 추후에 다른사람이 사용자에게 보여주는 부분을 수정할때 한번에 클래스를 찾고 유지보수 할 수 있을것 같아.

junuuuuuu commented on src/Car.java:
```diff
@@ -0,0 +1,35 @@
+    String name;
+    private String runCount = "";     //몇 번 움직였는지 상황
+
+    Car(String name) throws IllegalArgumentException {
+        while(true){
+            if(name.length()>=NAME_LENGTH_THRESHOLD) {
+                throw new IllegalArgumentException();
+            } else {
+                this.name = name;
+                break;
+            }
+        }
+    }
```
내부에서 이름을 반복적으로 입력받도록 수행해두었는데 이 경우에도 Input을 담당하는 클래스를 만들고 거기서 입력받으면 어떨까 싶어.

그리고 생성자에 while문이 들어갈 필요가 없어보이는것 같아

junuuuuuu commented on src/Finish.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * 게임종류 후 우승자를 찾는 클래스
+ */
+public class Finish {
+    public static void winner(ArrayList<Car> player) {
+        //runCount가 가장 높은 값 찾기
+        int winnerRunCount = 0;
+        for(int i=0; i<player.size(); i++) {
+            if(winnerRunCount < player.get(i).getRunCount()) {
+                winnerRunCount = player.get(i).getRunCount();
+            }
+        }
+
+        //runCount 가장 높은 값 출력
+        String winnerName = ""; //우승자 출력할 문자열
+        for(int i=0; i<player.size(); i++) {
+            if(player.get(i).getRunCount() == winnerRunCount) {
+                winnerName = winnerName.concat(player.get(i).name+",");
+            }
+        }
+
+        //마지막에 붙은 쉼표값 제거
+        if(winnerName.endsWith(",")){
+            winnerName = winnerName.substring(0,winnerName.length()-1);
+        };
+        System.out.println("최종 우승자: "+winnerName);
+    }
+}
```
Finish 클래스명보다 FindWinner라는 이름의 클래스를 가지면 어떨까 싶어, 그리고 for문을 작성하여 코드를 작성했는데 Java 8의 Stream을 활용하게되면 조금 더 가독성있는 프로그래밍이 가능할것 같아

junuuuuuu commented on src/Car.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Car {
+    final int NAME_LENGTH_THRESHOLD = 6;
+    Scanner input = new Scanner(System.in);
```
사용하지 않는 부분인것 같아 제거하면 될듯!

Weekndd committed:
Refactor

1. Finish 클래스명을 Findwinner로 수정.
2. Findwinner 클래스에 있는 For문을 Stream으로 수정해서 활용.
3. Input클래스를 만들어서 생성자에서 하던 예외처리를 분리.
4. Input클래스에 checkPlayerName으로 예외처리를 담당하는 메서드 생성.
5. View클래스를 만들어서 기존의 runCount는 Car클래스에서 숫자로 관리하고 화면 출력부분을 따로 분리.
6. RacingCarTest(테스트 코드)에 5글자 이상일 시 예외처리와 FindWinner클래스의 우승자를 출력하는 테스트작성
(예외처리가 정상적으로 돌아가는지 확인하는 테스트는 예외처리가 잘 되면 예외처리가 잘된다는 메시지와 함께 예외가 터짐..이게 맞나?)

Weekndd committed:
Delete Car.java

mis-upload

Weekndd committed:
Delete FindWinner.java

mis-upload

Weekndd committed:
Delete Input.java

mis-upload

Weekndd committed:
Delete Main.java

mis-upload

Weekndd committed:
Delete RacingCarTest.java

mis-upload

Weekndd committed:
Delete Ready.java

mis-upload

Weekndd committed:
Delete RandomNumberGenerater.java

mis-upload

Weekndd committed:
Delete Start.java

mis-upload

Weekndd committed:
Delete View.java

mis-upload

Weekndd committed:
Refactor

1. Finish 클래스명을 Findwinner로 수정.
2. Findwinner 클래스에 있는 For문을 Stream으로 수정해서 활용.
3. Input클래스를 만들어서 생성자에서 하던 예외처리를 분리.
4. Input클래스에 checkPlayerName으로 예외처리를 담당하는 메서드 생성.
5. View클래스를 만들어서 기존의 runCount는 Car클래스에서 숫자로 관리하고 화면 출력부분을 따로 분리.
6. RacingCarTest(테스트 코드)에 5글자 이상일 시 예외처리와 FindWinner클래스의 우승자를 출력하는 테스트작성
(예외처리가 정상적으로 돌아가는지 확인하는 테스트는 예외처리가 잘 되면 예외처리가 잘된다는 메시지와 함께 예외가 터짐..이게 맞나?)

Weekndd committed:
Delete Finish.java

Rename to FindWinner.java

junuuuuuu commented:
굿~ 일단 머지한번 해도 될듯
